File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {

    List<String> res = new ArrayList();
    int total;

    public List<String> generateParenthesis(int n) {
        // dynamic programming
        total = n;
        dfs(0, 0, "");
        return res;
    }


    private void dfs(int open, int closed, String str) {
        if(open == closed && open == total) {
            res.add(str);
            return;
        }

        if(open < total) {
            dfs(open+1, closed, str+"(");
        }

        if(closed < open) {
            dfs(open, closed+1, str+")");
        }
    }
}

File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        # stack = []
        res = []

        def dfs(open_count: int, close_count: int, curr: str):
            if open_count == close_count == n:
                res.append(curr)
                return
            
            if open_count < n:
                dfs(open_count+1, close_count, curr+"(")
            
            if close_count < open_count:
                dfs(open_count, close_count+1, curr+")")
        
        dfs(0,0,"")
        return res
        